modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java

```diff
@@ -856,335 +886,218 @@ private CompletableFuture<Object> processSingleEntryAction(ReadWriteSingleRowRep
-                                                .thenApply(tblLock -> null));
-                    }
+    private CompletableFuture<?> takePutLockOnIndexes(BinaryRow tableRow, RowId rowId, UUID txId) {
+        List<IndexLocker> indexes = indexesLockers.get();
 
-                    return CompletableFuture.completedFuture(rowId);
-                });
+        if (nullOrEmpty(indexes)) {
+            return CompletableFuture.completedFuture(null);
+        }
+
+        CompletableFuture<?>[] locks = new CompletableFuture[indexes.size()];
+        int idx = 0;
+
+        for (IndexLocker locker : indexes) {
+            locks[idx++] = locker.locksForInsert(txId, tableRow, rowId);
+        }
+
+        return CompletableFuture.allOf(locks);
     }
 
     /**
      * Takes all required locks on a key, before deleting the value.
      *
-     * @param searchKey Key to search.
-     * @param searchRow Row to remove.
-     * @param indexId   Index id.
      * @param txId      Transaction id.
      * @return Future completes with {@link RowId} or {@code null} if there is no value for remove.
      */
-    private CompletableFuture<RowId> takeLocksForDeleteExact(ByteBuffer searchKey, BinaryRow searchRow, UUID indexId, UUID txId) {
-        return lockManager.acquire(txId, new LockKey(indexId, searchKey), LockMode.X).thenCompose(idxLock -> { // Index X lock
-            RowId rowId = rowIdByKey(indexId, searchKey);
-
-            return lockManager.acquire(txId, new LockKey(tableId), LockMode.IX) // IX lock on table
-                    .thenCompose(tblLock -> {
-                        CompletableFuture<RowId> rowLockFut;
-
-                        if (rowId != null) {
-                            rowLockFut = lockManager.acquire(txId, new LockKey(tableId, rowId), LockMode.S) // S lock on RowId
-                                    .thenCompose(sharedRowLock -> {
-                                        BinaryRow curVal = resolveReadResult(mvDataStorage.read(rowId, HybridTimestamp.MAX_VALUE), txId);
-
-                                        if (equalValues(curVal, searchRow)) {
-                                            return lockManager.acquire(txId, new LockKey(tableId, rowId),
-                                                            LockMode.X) // X lock on RowId
-                                                    .thenApply(exclusiveRowLock -> rowId);
-                                        }
-
-                                        return CompletableFuture.completedFuture(null);
-                                    });
-                        } else {
-                            rowLockFut = CompletableFuture.completedFuture(null);
-                        }
+    private CompletableFuture<RowId> takeLocksForDeleteExact(BinaryRow expectedRow, RowId rowId, BinaryRow actualRow, UUID txId) {
+        return lockManager.acquire(txId, new LockKey(tableId, expectedRow.keySlice()), LockMode.X)  // Index X lock
```

Comment:
I'd rather expect index locks to be acquired wihtin index wrapper. Same for takeLocksForDelete, etc.

Reply:
fixed